docs/proposals/timeouts.md

```diff
@@ -0,0 +1,256 @@
+
+OutboundTimeout policy is a connection policy so it will be applied to the pair of service (A and B in the example). 
+The idea of the policy is to instruct Service A how to request Service B. Example:
+
+```yaml
+type: OutboundTimeout
+mesh: default
+name: default-timeouts-srv-B
+sources:
+- match:
+    kuma.io/service: '*'
+destinations:
+- match:
+    kuma.io/service: 'srv-B'
+conf:
+  # connectTimeout defines time to establish connection, 'connect_timeout' on Cluster [1]
+  connectTimeout: 10s 
+  tcp:
+    # 'idle_timeout' on TCPProxy [2]
+    idleTimeout: 1h
+  http:
+    # 'timeout' on Route [7]
+    requestTimeout: 5s
+    # 'idle_timeout' on Cluster [3]
+    idleTimeout: 1h 
+  grpc:
+    # 'stream_idle_timeout' on HttpConnectionManager [6]
+    streamIdleTimeout: 5m
+    # 'max_stream_duration' on Cluster [5]
+    maxStreamDuration: 30m
+``` 
+
+As an owner of Service B you have some SLA regarding the upper boundaries of the request/response time, so you can apply
+this `default-timeouts-srv-B` and every service which is trying to consume Service B will have those timeouts by default.
+
+At the same time if you are an owner of Service A you can override this policy for your service and put different timeout 
+values. From the Service B perspective `default-timeouts-srv-B` policy has an advisory nature, every service can override it,
+`OutboundTimeout` policy is not purposed to protect Service B. In order to protect Service B from abusive clients with long
+requests it’s better to use `InboundTimeouts`.
+
+### InboundTimeout
+
+Unlike the `OutboundTimeout` this policy is matched for the Dataplane rather than pair of services (applied only to the Dataplanes of the Service B). `InboundTimeout` has the same configuration, but values are mapped to the different Envoy timeouts:
+
+```yaml
+type: InboundTimeout
+mesh: default
+name: timeouts-srv-B
+selectors:
+  - match:
+      kuma.io/service: srv-B
+conf:
+  connectTimeout: 10s # 'transport_socket_connect_timeout' on FilterChain [12]
+  tcp:
+    # 'idle_timeout' on TCPProxy [2]
+    idleTimeout: 1h
+  http:
+    # 'request_timeout' on HttpConnectionManager [4]
+    requestTimeout: 10s
+    # 'request_headers_timeout' on HttpConnectionManager [11]
+    requestHeadersTimeout: 5s
+    # 'idle_timeout' on HttpConnectionManager [3]
+    idleTimeout: 1h
+  grpc:
+    # 'stream_idle_timeout' on HttpConnectionManager [6]
+    streamIdleTimeout: 5s
+    # 'max_stream_duration' on HttpConnectionManager [5]
+    maxStreamDuration: 30m
+``` 
+
+This policy aims to protect your service from the long living requests, if the service is not designed to support them.
+
+### Uncovered timeouts
+
+There are 3 uncovered timeouts - [8], [9], [10].
+
+[8] and [10] allows to override existing timeouts on per-route basis, probably it makes sense to implement them in L7 
+TrafficRoutes.
+
+[9] - probably makes sense to implement in the scope of Retry policy. 
+
```

Comment:
I like the symmetry of the naming of policies, but I'm not fully on board with it.
I think that in 80% of cases users will care about `OutboundTimeout` and stick to the default `InboundTimeout`.

I'd rename `OutboundTimeout` to `Timeout`, but then I have no good idea on another name for `InboundTimeout`.

What do you think of embedding `InboundTimeout` into `Bulkhead` policy? I think `Bulkhead` will be designed for exactly this use case - protecting your service from the attacks.

Reply:
Yes, I actually also had this in mind `InboundTimeout` and `Bulkhead` are kind of about the same, but we don't have an opinion about `Bulkhead` itself yet. Maybe we have to include `Bulkhead` into `CircuitBreaker`. If so it's better to keep `InboundTimeout` separate. 

I agree to rename `OutboundTimeout` to `Timeout` since it's expected to be more actively used than `InboundTimeout`.